book-content/chapters/14-configuring-typescript.md

```diff
@@ -0,0 +1,1024 @@
+{
+  "compilerOptions": {
+    "declarationMap": true,
+    "sourceMap": true
+  }
+}
+```
+
+With this option in place, the TypeScript compiler will generate `.d.ts.map` files alongside the `.d.ts` files. Now, when you `CMD + click` on an import in `my-app`, you'll be taken to the original TypeScript source file in `my-library`.
+
+This is less useful when your library is on `npm`, unless you also ship your source files - but that's a little outside the scope of this book. In a monorepo, however, declaration maps are a great quality-of-life improvement.
+
+## `jsx`
+
+TypeScript has built-in support for transpiling JSX syntax, which is a syntax extension for JavaScript that allows you to write HTML-like code in your JavaScript files. In TypeScript, these need to be written in files with a `.tsx` extension:
+
+```tsx
+// Component.tsx
+const Component = () => {
+  return <div />;
+};
+```
+
+The `jsx` option tells TypeScript how to handle JSX syntax, and has five possible values. The most common are `preserve`, `react`, and `react-jsx`. Here's what each of them does:
+
+- `preserve`: Keeps JSX syntax as-is.
+- `react`: Transforms JSX into `React.createElement` calls. Useful for React 16 and earlier.
+- `react-jsx`: Transforms JSX into `_jsx` calls, and automatically imports from `react/jsx-runtime`. Useful for React 17 and later.
+
+## Managing Multiple TypeScript Configurations
+
+As projects grow in size and complexity, it's common to have different environments or targets within the same project.
+
+For example, your single repo might include both a client-side application and a server-side API, each with different requirements and configurations.
+
+This means you might want different `tsconfig.json` settings for different parts of your project. In this section, we'll look at how multiple `tsconfig.json` files can be composed together.
+
+### How TypeScript Finds `tsconfig.json`
+
+In a project with multiple `tsconfig.json` files, your IDE will need to know which one to use for each file. It determines which `tsconfig.json` to use by looking for the closest one to the current `.ts` file in question.
+
+For example, given this file structure:
+
+```
+project
+  ├── client
+  │   └── tsconfig.json
+  ├── server
+  │   └── tsconfig.json
+  └── tsconfig.json
+```
+
+A file inside the `client` directory will use the `client/tsconfig.json` file, while a file inside the `server` directory will use the `server/tsconfig.json` file. Anything inside the `project` directory that isn't in `client` or `server` will use the `tsconfig.json` file in the root of the project.
+
+This means that `client/tsconfig.json` can contain settings specific to the client-side application, such as adding the `dom` types:
+
+```json
+{
+  "compilerOptions": {
+    // ...other options
+    "lib": ["es2022", "dom", "dom.iterable"]
+  }
+}
+```
+
+But `server/tsconfig.json` can contain settings specific to the server-side application, such as removing the `dom` types:
```

Comment:
this is also a good way of structuring things for src and test dirs... although usually people choose the easier option to just shove everything into a single tsconfig